core/src/Torch/Raw/Tensor/Generic.hs

```diff
@@ -199,53 +205,176 @@ getDynamicDim
 fillZeros :: (THTensorMath t, THTensor t, Num (HaskReal t)) => Ptr t -> IO (Ptr t)
 fillZeros t = c_fill t 0 >> pure t
 
+-- Display stuff
+minLogScale :: Int
+minLogScale =
+    ceiling $ logBase 10 p
+  where
+    minPositive = NL.minValue * NL.epsilon  -- get smallest denormal
+    p = if (minPositive == 0) then NL.minValue else minPositive
+
+numberFormat :: RealFloat a => Int -> Int -> [a] -> (String, a, Int)
+numberFormat precision minSz xs = (format, scale, sz)
+  where
+    listParams :: (RealFloat a) => [a] -> (a, a, Bool, Bool, Bool)
+    listParams z =
+      f (NL.infinity, -NL.infinity, True, False, False) z
+      where
+        f z [] = z
+        f (minAcc, maxAcc, intModeAcc, hasInvalidAcc, hasNegInfAcc) (x:xs) = 
```

Comment:
I think this can be cleaned up with a `foldMap` over some newtype deriving Monoid which encodes the rules you are introducing. The cascading if statements take a while to process which triggers a memory of degoes' [_Destroy All Ifs_](http://degoes.net/articles/destroy-all-ifs) post.

Reply:
Scanning the rest of this function and re-scanning _Destroy All Ifs_, it also looks like you could replace some of those bools with Sum-type ADTs. I think this would also warrant a new module for printing which can be exported by this one. Maybe `Torch.Raw.Tensor.Printing` for now and we can discuss later if we need a `Torch.Generic` parent.